src/v/cluster/tm_stm.cc

```diff
@@ -337,38 +337,74 @@ ss::future<tm_stm::op_status> tm_stm::register_new_producer(
     co_return tm_stm::op_status::success;
 }
 
-bool tm_stm::add_partitions(
+ss::future<tm_stm::op_status> tm_stm::add_partitions(
   kafka::transactional_id tx_id,
   std::vector<tm_transaction::tx_partition> partitions) {
     auto ptx = _mem_txes.find(tx_id);
     if (ptx == _mem_txes.end()) {
-        return false;
+        co_return tm_stm::op_status::unknown;
     }
     if (ptx->second.status != tm_transaction::tx_status::ongoing) {
-        return false;
+        co_return tm_stm::op_status::unknown;
     }
+    bool just_started = ptx->second.partitions.size() == 0
+                        && ptx->second.groups.size() == 0;
+
+    if (just_started) {
+        tm_transaction tx = ptx->second;
+        for (auto& partition : partitions) {
+            tx.partitions.push_back(partition);
+        }
+        tx.last_update_ts = clock_type::now();
+        auto r = co_await update_tx(tx, tx.etag);
+
+        if (!r.has_value()) {
+            co_return tm_stm::op_status::unknown;
+        }
+        _mem_txes[tx_id] = tx;
```

Comment:
q: Since add_partitions can happen in multiple RPCs, how does this handle the following sequence?

```
-- mem_state = log_state = {}
add_partitions(tx1, <ntp1, ntp2>) - tm_stm on host1
 -- mem_state -> tx1 =<ntp1, ntp2>
 -- log_state -> tx1 = <ntp1, ntp2>

host1 crash
tm_stm recovers on host2

-- log_state =  tx1 = <ntp1, ntp2>
-- mem_state = {}
add_partitions(tx1, <ntp3, ntp4> 
  -- mem_state = tx1 = <ntp3, ntp4>
  -- log_state = tx1 = <ntp3, ntp4> --> gets overwritten?
```

We get a divergence, no?

Reply:
it fails on host2 because tx's etag from log_state doesn't match the current term